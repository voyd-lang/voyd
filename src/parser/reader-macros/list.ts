import { ReaderMacro } from "./types.js";

export const listReader: ReaderMacro = {
  match: (t) => t.value === "(" || t.value === ")",
  macro: (file, { token, reader }) => {
    if (token.value === "(") {
      const list = reader(file, ")");
      list.setAttribute("tuple?", true);
      return list;
    }

    return undefined; // discard closing paren
  },
};
