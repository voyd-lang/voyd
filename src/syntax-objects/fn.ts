import type { Expr } from "./expr.js";
import { Identifier } from "./identifier.js";
import { ChildList } from "./lib/child-list.js";
import { Child } from "./lib/child.js";
import { ScopedNamedEntity, ScopedNamedEntityOpts } from "./named-entity.js";
import { Parameter } from "./parameter.js";
import { FnType, Type, selfType } from "./types.js";
import { Variable } from "./variable.js";

export class Fn extends ScopedNamedEntity {
  readonly syntaxType = "fn";
  readonly #parameters = new ChildList<Parameter>([], this);
  readonly #body = new Child<Expr | undefined>(undefined, this);
  readonly #returnTypeExpr = new Child<Expr | undefined>(undefined, this);
  readonly #genericInstances = new ChildList<Fn>([], this);
  #typeParams = new ChildList<Identifier>([], this);
  variables: Variable[] = [];
  returnType?: Type; // When a function has generics, resolved versions of the functions go here
  inferredReturnType?: Type;
  annotatedReturnType?: Type;
  appliedTypeArgs?: Type[] = [];
  typesResolved?: boolean;
  #iteration = 0;

  constructor(
    opts: ScopedNamedEntityOpts & {
      returnTypeExpr?: Expr;
      variables?: Variable[];
      parameters?: Parameter[];
      typeParameters?: Identifier[];
      body?: Expr;
    }
  ) {
    super(opts);
    this.#parameters.push(...(opts.parameters ?? []));
    this.#typeParams.push(...(opts.typeParameters ?? []));
    this.returnTypeExpr = opts.returnTypeExpr;
    this.variables = opts.variables ?? [];
    this.body = opts.body;
  }

  get body() {
    return this.#body.value;
  }

  set body(body: Expr | undefined) {
    this.#body.value = body;
  }

  get parameters() {
    return this.#parameters.toArray();
  }

  get returnTypeExpr() {
    return this.#returnTypeExpr.value;
  }

  set returnTypeExpr(returnTypeExpr: Expr | undefined) {
    this.#returnTypeExpr.value = returnTypeExpr;
  }

  get genericInstances() {
    const instances = this.#genericInstances.toArray();
    return !instances.length ? undefined : instances;
  }

  get typeParameters() {
    const params = this.#typeParams.toArray();
    return !params.length ? undefined : params;
  }

  set typeParameters(params: Identifier[] | undefined) {
    this.#typeParams = new ChildList(params ?? [], this);
  }

  // Register a version of this function with resolved generics
  registerGenericInstance(fn: Fn) {
    this.#genericInstances.push(fn);
  }

  getNameStr(): string {
    return this.name.value;
  }

  getType(): FnType {
    return new FnType({
      ...super.getCloneOpts(this.parent),
      parameters: this.parameters,
      returnType: this.getReturnType(),
    });
  }

  getIndexOfParameter(parameter: Parameter) {
    const index = this.parameters.findIndex((p) => p.id === parameter.id);
    if (index < 0) {
      throw new Error(`Parameter ${parameter} not registered with fn ${this}`);
    }
    return index;
  }

  getIndexOfVariable(variable: Variable) {
    const index = this.variables.findIndex((v) => v.id === variable.id);

    if (index < 0) {
      const newIndex = this.variables.push(variable) - 1;
      return newIndex + this.parameters.length;
    }

    return index + this.parameters.length;
  }

  getReturnType(): Type {
    if (this.returnType) {
      return this.returnType;
    }

    throw new Error(
      `Return type not yet resolved for fn ${this.name} at ${this.location}`
    );
  }

  toString() {
    return this.id;
  }

  clone(parent?: Expr | undefined): Fn {
    // Don't clone generic instances
    return new Fn({
      ...super.getCloneOpts(parent),
      id: `${this.id}#${this.#iteration++}`,
      returnTypeExpr: this.returnTypeExpr?.clone(),
      parameters: this.#parameters.clone(),
      typeParameters: this.#typeParams.clone(),
      body: this.body?.clone(),
    });
  }

  toJSON(): unknown {
    return [
      "fn",
      this.id,
      ["parameters", ...this.parameters],
      ["type-parameters", ...(this.#typeParams.toArray() ?? [])],
      ["return-type", this.#returnTypeExpr?.toJSON() ?? null],
      this.body,
    ];
  }
}
