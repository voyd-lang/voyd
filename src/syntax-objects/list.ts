import { Expr } from "./expr.js";
import { Float } from "./float.js";
import { getIdStr } from "./lib/get-id-str.js";
import { Id, Identifier } from "./identifier.js";
import { Int } from "./int.js";
import { SourceLocation, Syntax, SyntaxMetadata } from "./syntax.js";
import { ChildList } from "./lib/child-list.js";

type ListOpts =
  | ListValue[]
  | (SyntaxMetadata & {
      value?: ListValue[] | List | ChildList<Expr>;
      dynamicLocation?: boolean;
    });

export class List extends Syntax {
  readonly syntaxType = "list";
  #store = new ChildList(undefined, this);
  dynamicLocation?: boolean;

  constructor(opts: ListOpts) {
    opts = Array.isArray(opts) ? { value: opts } : opts;
    super(opts);
    const value = opts.value;
    this.dynamicLocation = opts.dynamicLocation;

    if (value instanceof Array) {
      value.forEach((v) => this.push(v));
    } else if (value instanceof List) {
      value.toArray().forEach((v) => this.push(v));
    }
  }

  get children() {
    return this.toArray();
  }

  get hasChildren() {
    return !!this.#store.length;
  }

  get length() {
    return this.#store.length;
  }

  at(index: number): Expr | undefined {
    return this.#store.at(index);
  }

  exprAt(index: number): Expr {
    const expr = this.#store.at(index);
    if (!expr) {
      throw new Error(`No expr at ${index}`);
    }
    return expr;
  }

  identifierAt(index: number): Identifier {
    const id = this.at(index);
    if (!id?.isIdentifier()) {
      throw new Error(`No identifier at index ${index}`);
    }
    return id;
  }

  optionalIdentifierAt(index: number): Identifier | undefined {
    const id = this.at(index);
    if (id?.isIdentifier()) return id;
  }

  listAt(index: number): List {
    const id = this.at(index);
    if (!id?.isList()) {
      throw new Error(`No list at index ${index}`);
    }
    return id;
  }

  set(index: number, expr: Expr | string) {
    const result = typeof expr === "string" ? Identifier.from(expr) : expr;
    this.#store.set(index, result);
    return this;
  }

  calls(fnId: Id, atIndex = 0) {
    return this.getIdStrAt(atIndex) === getIdStr(fnId);
  }

  getIdStrAt(index: number): string | undefined {
    const v = this.at(index);
    return v?.isIdentifier() || v?.isStringLiteral() ? v.value : undefined;
  }

  consume(): Expr {
    return this.#store.consume();
  }

  first(): Expr | undefined {
    return this.#store.at(0);
  }

  last(): Expr | undefined {
    return this.#store.at(-1);
  }

  /** Returns all but the first element in an array */
  argsArray(): Expr[] {
    return this.#store.toArray().slice(1);
  }

  pop(): Expr | undefined {
    return this.#store.pop();
  }

  push(...expr: ListValue[]) {
    expr.forEach((ex) => {
      if (typeof ex === "string") {
        this.#store.push(new Identifier({ value: ex, parent: this }));
        return;
      }

      if (typeof ex === "number" && Number.isInteger(ex)) {
        this.#store.push(new Int({ value: ex, parent: this }));
        return;
      }

      if (typeof ex === "number") {
        this.#store.push(new Float({ value: ex, parent: this }));
        return;
      }

      if (ex instanceof Array) {
        this.push(new List({ value: ex, parent: this }));
        return;
      }

      if (ex.syntaxType === "nop") {
        return;
      }

      if (this.dynamicLocation) {
        this.updateLocationFrom(ex);
      }

      this.#store.push(ex);
    });

    return this;
  }

  private updateLocationFrom(ex: Expr) {
    if (!ex.location) return;

    if (!this.location) {
      this.location = ex.location.clone();
    }

    if (this.location.startIndex > ex.location.startIndex) {
      const nl = ex.location.clone();
      nl.setEndToEndOf(this.location);
      this.location = nl;
    }

    if (
      !this.location.endColumn ||
      this.location.endIndex < ex.location.endIndex
    ) {
      this.location?.setEndToEndOf(ex.location);
    }
  }

  findIndex(cb: (expr: Expr) => boolean) {
    return this.toArray().findIndex(cb);
  }

  insert(expr: Expr | string, at = 0) {
    const result = typeof expr === "string" ? Identifier.from(expr) : expr;
    this.#store.insert(result, at);
    return this;
  }

  remove(index: number, count = 1) {
    this.#store.remove(index, count);
    return this;
  }

  each(fn: (expr: Expr, index: number, array: Expr[]) => void): List {
    this.toArray().forEach(fn);
    return this;
  }

  map(fn: (expr: Expr, index: number, array: Expr[]) => Expr): List {
    return new List({
      ...super.getCloneOpts(),
      value: this.toArray().map(fn),
    });
  }

  flatMap(
    fn: (expr: Expr, index: number, array: Expr[]) => Expr | Expr[]
  ): List {
    return new List({
      ...super.getCloneOpts(),
      value: this.toArray().flatMap(fn),
    });
  }

  reduce(
    fn: (acc: List, expr: Expr, index: number, array: Expr[]) => List
  ): List {
    return this.toArray().reduce(fn, new List({ ...super.getCloneOpts() }));
  }

  slice(start?: number, end?: number): List {
    return new List({
      ...super.getCloneOpts(),
      value: this.#store.slice(start, end),
    });
  }

  sliceAsArray(start?: number, end?: number) {
    return this.children.slice(start, end);
  }

  /**
   * Returns the expression for a labeled argument, eg. for the list
   * `[hello, [:, world, 1]]` – which represents `hello(world: 1)` –
   * `labeledArg("world")` will return the expression `1`.
   *
   * If the label is not found, `undefined` is returned.
   */
  optionalLabeledArg(label: string): Expr | undefined {
    for (const expr of this.#store.toArray()) {
      if (!expr.isCall()) continue;
      if (!expr.calls(":")) continue;

      const labelId = expr.argAt(0);
      if (!labelId?.isIdentifier()) continue;

      if (labelId.value === label) {
        return expr.argAt(1);
      }
    }

    return undefined;
  }

  /** Asserts that a labeled argument exists and returns its expression. */
  labeledArg(label: string): Expr {
    const expr = this.optionalLabeledArg(label);
    if (!expr) throw new Error(`Labeled argument '${label}' not found`);
    return expr;
  }

  /**
   * Returns `true` when the list contains a labeled argument with the provided
   * name.
   */
  hasLabeledArg(label: string): boolean {
    return this.optionalLabeledArg(label) !== undefined;
  }

  /**
   * Builds and returns a map of all labeled arguments contained in the list.
   */
  getLabeledArgs(): Map<string, Expr> {
    const map = new Map<string, Expr>();
    for (const expr of this.#store.toArray()) {
      if (!expr.isCall() || !expr.calls(":")) continue;
      const labelId = expr.argAt(0);
      if (!labelId?.isIdentifier()) continue;
      const valueExpr = expr.argAt(1);
      if (valueExpr) map.set(labelId.value, valueExpr);
    }
    return map;
  }

  toArray(): Expr[] {
    return this.#store.toArray();
  }

  toJSON() {
    return this.toArray();
  }

  /** Clones should return a deep copy of all expressions except for type expressions */
  clone(parent?: Expr): List {
    return new List({
      ...super.getCloneOpts(parent),
      value: this.#store.clone(),
    });
  }
}

export type ListValue = Expr | string | number | ListValue[];
