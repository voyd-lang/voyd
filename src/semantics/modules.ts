import { ParsedModule, stdPath } from "../parser/index.js";
import { List } from "../syntax-objects/list.js";
import { RootModule, VoydModule } from "../syntax-objects/module.js";

/** Registers submodules of a parsed module for future import resolution */
export const registerModules = (
  opts: ParsedModule,
  rootModule: VoydModule = new RootModule({})
): VoydModule => {
  const { srcPath, files } = opts;

  for (const [filePath, file] of Object.entries(files)) {
    const resolvedPath = filePathToModulePath(
      filePath,
      srcPath ?? opts.indexPath
    );

    const parsedPath = resolvedPath.split("/").filter(Boolean);

    registerModule({
      path: parsedPath,
      parentModule: rootModule,
      ast: file.slice(1), // Skip the first element (ast)
      isIndex: filePath === opts.indexPath,
    });
  }

  return rootModule;
};

const registerModule = ({
  path,
  parentModule,
  ast,
  isIndex,
}: {
  path: string[];
  parentModule: VoydModule;
  ast: List;
  isIndex?: boolean;
}): VoydModule | undefined => {
  const [name, ...rest] = path;

  if (!name) return;

  const existingModule = parentModule.resolveEntity(name);

  if (existingModule && !existingModule.isModule()) {
    throw new Error(
      `Cannot register module ${name} because it is already registered as ${existingModule.syntaxType}`
    );
  }

  if (!existingModule && (name === "index" || name === "mod")) {
    parentModule.value = ast.toArray();
    registerDefaultImports(parentModule);
    parentModule.getAllEntities().forEach((e) => {
      if (e.isModule()) parentModule.push(e);
    });
    return;
  }

  const module =
    existingModule ??
    new VoydModule({
      ...(!rest.length ? { ...ast.metadata, value: ast.toArray() } : {}),
      name,
      isIndex,
    });

  if (!existingModule) {
    parentModule.push(module);
    registerDefaultImports(module);
  }

  if (!existingModule && (module.name.is("src") || module.name.is("std"))) {
    parentModule.registerExport(module);
  }

  if (existingModule && !rest.length) {
    module.unshift(...ast.toArray().reverse());
  }

  if (!rest.length) return;

  return registerModule({ path: rest, parentModule: module, ast });
};

const filePathToModulePath = (filePath: string, srcPath: string) => {
  let finalPath = filePath.startsWith(stdPath)
    ? filePath.replace(stdPath, "std")
    : filePath;

  finalPath = finalPath.startsWith(srcPath)
    ? finalPath.replace(srcPath, "src")
    : finalPath;

  finalPath = finalPath.replace(".voyd", "");

  return finalPath;
};

const registerDefaultImports = (module: VoydModule) => {
  module.unshift(new List(["use", ["::", "root", "all"]]));
  const mod = module.resolveModule("std");
  if (mod) return;
  module.unshift(new List(["use", ["::", "std", "all"]]));
};
