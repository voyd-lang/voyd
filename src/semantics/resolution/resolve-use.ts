import { Call } from "../../syntax-objects/call.js";
import { Expr } from "../../syntax-objects/expr.js";
import { List } from "../../syntax-objects/list.js";
import { VoydModule } from "../../syntax-objects/module.js";
import { NamedEntity } from "../../syntax-objects/named-entity.js";
import { Use, UseEntities } from "../../syntax-objects/use.js";
import { resolveModule, resolveEntities } from "./resolve-entities.js";

export type ModulePass = (mod: VoydModule) => VoydModule;

export const resolveUse = (use: Use, runPass?: ModulePass) => {
  const path = use.path;

  const entities = resolveModulePath(path, runPass);
  entities.forEach((e) => use.parentModule?.registerEntity(e.e, e.alias));
  use.entities = entities;
  return use;
};

export const resolveModulePath = (
  path: Expr,
  runPass?: ModulePass
): { e: NamedEntity; alias?: string }[] => {
  if (path.isIdentifier()) {
    const e = path.resolve();
    return e ? [{ e }] : [];
  }

  if (!path.isCall() && !path.isList()) {
    throw new Error("Invalid path statement");
  }

  if (!path.calls("::")) {
    throw new Error(`Invalid use statement ${path}`);
  }

  const [left, right] = path.argsArray();

  const resolvedModule =
    left?.isList() || left?.isCall()
      ? resolveModulePath(left, runPass)[0]?.e
      : left?.isIdentifier()
      ? left.resolveModule(left)
      : undefined;

  if (
    !resolvedModule ||
    resolvedModule instanceof Array ||
    !resolvedModule.isModule()
  ) {
    throw new Error(
      `Invalid use statement, not a module ${JSON.stringify(
        path,
        null,
        2
      )} at ${path.location}`
    );
  }

  const module = runPass ? runPass(resolvedModule) : resolvedModule;

  if ((right.isCall() || right.isList()) && right.calls("object")) {
    return resolveObjectPath(right, module);
  }

  if (!right?.isIdentifier()) {
    if (resolvedModule.phase < 3) return []; // Ignore unresolved entities in macro phase
    throw new Error(`Invalid use statement, expected identifier, got ${right}`);
  }

  if (right?.is("all")) {
    return module.getAllExports().map((e) => ({ e }));
  }

  const entities = module.resolveExport(right);
  if (!entities.length) {
    if (resolvedModule.phase < 3) return []; // Ignore unresolved entities in macro phase
    throw new Error(
      `Invalid use statement, entity ${right} is not exported at ${right.location}`
    );
  }

  return entities.map((e) => ({ e }));
};

const resolveObjectPath = (path: Call | List, module: VoydModule) => {
  const entities: UseEntities = [];

  const imports = path.argsArray();
  for (const imp of imports) {
    if (
      (imp.isCall() || imp.isList()) &&
      (imp.calls(":") || imp.calls("as"))
    ) {
      const args = imp.argsArray();
      const entityId = args.at(0);
      const alias = args.at(1);
      if (!entityId?.isIdentifier() || !alias?.isIdentifier()) {
        if (module.phase < 3) continue; // Ignore unresolved entities in macro phase
        throw new Error(
          `Invalid use statement, expected identifiers, got ${imp}`
        );
      }

      if (entityId.is("self")) {
        entities.push({ e: module, alias: alias?.value });
        continue;
      }

      const resolvedEntities = module.resolveExport(entityId);
      if (!resolvedEntities.length) {
        if (module.phase < 3) continue; // Ignore unresolved entities in macro phase
        throw new Error(
          `Invalid use statement, entity ${entityId} is not exported at ${entityId.location}`
        );
      }

      entities.push(
        ...resolvedEntities.map((e) => ({ e, alias: alias?.value }))
      );
      continue;
    }

    if (!imp.isIdentifier()) {
      throw new Error(`Invalid use statement, expected identifier, got ${imp}`);
    }

    if (imp.is("self")) {
      entities.push({ e: module });
      continue;
    }

    const resolvedEntities = module.resolveExport(imp);
    if (!resolvedEntities.length) {
      if (module.phase < 3) continue; // Ignore unresolved entities in macro phase
      throw new Error(
        `Invalid use statement, entity ${imp} is not exported at ${imp.location}`
      );
    }

    entities.push(...resolvedEntities.map((e) => ({ e })));
  }

  return entities;
};

export const resolveExport = (call: Call) => {
  const block = call.argAt(0);
  if (!block?.isBlock()) return call;

  const entities = block.body.map(resolveEntities);
  registerExports(call, entities, resolveModule);

  return call;
};

export const registerExports = (
  exportExpr: Expr,
  entities: (Expr | NamedEntity | { e: NamedEntity; alias?: string })[],
  pass?: ModulePass
) => {
  entities.forEach((e) => {
    if ("e" in e) {
      registerExport(exportExpr, e.e, e.alias);
      return;
    }

    if (e.isUse()) {
      e.entities.forEach((e) => registerExport(exportExpr, e.e, e.alias));
      return;
    }

    if ((e.isCall() || e.isList()) && e.calls("mod")) {
      registerExports(exportExpr, resolveModulePath(e.argsArray()[0], pass));
      return;
    }

    if (e instanceof NamedEntity) {
      registerExport(exportExpr, e);
      if (!e.parentImpl) e.parentModule?.registerEntity(e);
    }
  });
};

const registerExport = (
  exportExpr: Expr,
  entity: NamedEntity,
  alias?: string
) => {
  const parent = exportExpr.parent;
  if (!parent) return;

  if (parent.isModule()) {
    parent.registerExport(entity, alias);
    return;
  }

  if (exportExpr.parentImpl && entity.isFn()) {
    exportExpr.parentImpl.registerExport(entity);
    return;
  }
};
