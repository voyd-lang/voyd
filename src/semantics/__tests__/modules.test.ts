import { registerModules } from "../modules.js";
import { List } from "../../syntax-objects/list.js";
import { stdPath } from "../../parser/index.js";
import { test } from "vitest";
import { RootModule, VoydModule } from "../../syntax-objects/module.js";
import { Fn } from "../../syntax-objects/fn.js";
import { resolveModulePath } from "../resolution/resolve-use.js";

test("module registration", (t) => {
  const result = registerModules(input);
  t.expect(result).toMatchSnapshot();
});

test("resolveModulePath handles import variants", (t) => {
  const root = new RootModule({});
  const std = new VoydModule({ name: "std", parent: root });
  const linearMem = new VoydModule({ name: "linear_memory", parent: std });

  const grow = new Fn({ name: "grow", parent: linearMem });
  const store = new Fn({ name: "store", parent: linearMem });
  const load = new Fn({ name: "load_i32", parent: linearMem });

  linearMem.registerExport(grow);
  linearMem.registerExport(store);
  linearMem.registerExport(load);

  std.registerEntity(linearMem);
  std.registerExport(linearMem);
  root.registerEntity(std);
  root.registerExport(std);

  const multiPath = new List([
    "::",
    new List(["::", "std", "linear_memory"]),
    new List([
      "object",
      new List([":", "grow", "grow_linear_mem"]),
      "store",
      "load_i32",
    ]),
  ]);
  multiPath.parent = root;
  const multiEntities = resolveModulePath(multiPath);
  t.expect(multiEntities.map((e) => e.e.name.value)).toEqual([
    "grow",
    "store",
    "load_i32",
  ]);
  t.expect(multiEntities[0].alias).toBe("grow_linear_mem");

  const singlePath = new List([
    "::",
    new List(["::", "std", "linear_memory"]),
    "grow",
  ]);
  singlePath.parent = root;
  const singleEntities = resolveModulePath(singlePath);
  t.expect(singleEntities[0].e).toBe(grow);

  const modulePath = new List(["::", "std", "linear_memory"]);
  modulePath.parent = root;
  const moduleEntities = resolveModulePath(modulePath);
  t.expect(moduleEntities[0].e).toBe(linearMem);
});

const input = {
  files: {
    "/Users/drew/projects/voyd/example.voyd": new List({
      value: [
        "ast",
        ["use", ["::", ["::", "std", "macros"], "all"]],
        [
          "fn",
          ["fib", [":", "n", "i32"]],
          "->",
          "i32",
          [
            "block",
            [
              "if",
              ["<=", "n", 1],
              [":", "then", ["block", "n"]],
              [
                ":",
                "else",
                [
                  "block",
                  ["+", ["fib", ["-", "n", 1]], ["fib", ["-", "n", 2]]],
                ],
              ],
            ],
          ],
        ],
        [
          "fn",
          ["main"],
          [
            "block",
            ["let", ["=", "x", ["+", 10, ["block", ["+", 20, 30]]]]],
            [
              "let",
              [
                "=",
                "y",
                [
                  "if",
                  [">", "x", 10],
                  [":", "then", ["block", 10]],
                  [":", "else", ["block", 20]],
                ],
              ],
            ],
            [
              "call",
              "this",
              "while",
              [
                "=>",
                [],
                [
                  "if",
                  [">", "x", 10],
                  [":", "then", ["block", ["-=", "x", 1]]],
                  [":", "else", ["block", ["+=", "x", 1]]],
                ],
              ],
            ],
            [
              "let",
              [
                "=",
                "n",
                [
                  "if",
                  [">", ["len", "args"], 1],
                  [
                    ":",
                    "then",
                    [
                      "block",
                      ["log", "console", ["string", "Hey there!"]],
                      ["unwrap", ["parseInt", ["at", "args", 1]]],
                    ],
                  ],
                  [":", "else", ["block", 10]],
                ],
              ],
            ],
            ["let", ["=", "x2", 10]],
            ["let", ["=", "z", ["nothing"]]],
            ["let", ["=", "test_spacing", ["fib", "n"]]],
            ["let", ["=", "result", ["fib", "n"]]],
          ],
        ],
      ],
    }),
    [`${stdPath}/deep/nested/hey.voyd`]: new List({
      value: [
        "ast",
        ["use", ["::", ["::", "super", "macros"], "all"]],
        [
          "pub",
          "fn",
          ["hey", [":", "src", "i32"], [":", "dest", "i32"]],
          "->",
          "i32",
          ["block"],
        ],
      ],
    }),
    [`${stdPath}/deep/nested.voyd`]: new List({
      value: [
        "ast",
        ["pub", ["use", ["::", "hey", "all"]]],
        [
          "pub",
          "fn",
          ["nested", [":", "src", "i32"], [":", "dest", "i32"]],
          "->",
          "i32",
          ["block"],
        ],
      ],
    }),
    [`${stdPath}/deep/mod.voyd`]: new List({
      value: [
        "ast",
        ["pub", ["use", ["::", "hey", "all"]]],
        [
          "pub",
          "fn",
          ["deeply", [":", "src", "i32"], [":", "dest", "i32"]],
          "->",
          "i32",
          ["block"],
        ],
      ],
    }),
    [`${stdPath}/memory.voyd`]: new List({
      value: [
        "ast",
        ["use", ["::", ["::", "super", "macros"], "all"]],
        ["global", "let", ["=", "header-size", 8]],
        ["global", "let", ["=", "size-index", 0]],
        ["global", "let", ["=", "type-index", 4]],
        ["global", "var", ["=", "stack-pointer", 0]],
        [
          "pub",
          "fn",
          ["copy", [":", "src", "i32"], [":", "dest", "i32"]],
          "->",
          "i32",
          [
            "block",
            [
              "bnr",
              ["memory", "copy", "voyd"],
              ["dest", "src", ["size", "src"]],
            ],
            "dest",
          ],
        ],
      ],
    }),
    [`${stdPath}/index.voyd`]: new List({
      value: ["ast", ["pub", ["use", ["::", "macros", "all"]]]],
    }),
  },
  srcPath: "/Users/drew/projects/voyd",
  indexPath: "/Users/drew/projects/voyd/index.voyd",
};
