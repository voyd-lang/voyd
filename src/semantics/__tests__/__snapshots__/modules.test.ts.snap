// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`module registration 1`] = `
[
  "module",
  "root",
  [
    "exports",
    [
      "src#377",
      "std#406",
    ],
  ],
  [
    [
      "module",
      "src",
      [
        "exports",
        [],
      ],
      [
        [
          "use",
          [
            "::",
            "std",
            "all",
          ],
        ],
        [
          "use",
          [
            "::",
            "root",
            "all",
          ],
        ],
        [
          "module",
          "example",
          [
            "exports",
            [],
          ],
          [
            [
              "use",
              [
                "::",
                "std",
                "all",
              ],
            ],
            [
              "use",
              [
                "::",
                "root",
                "all",
              ],
            ],
            [
              "use",
              [
                "::",
                [
                  "::",
                  "std",
                  "macros",
                ],
                "all",
              ],
            ],
            [
              "fn",
              [
                "fib",
                [
                  ":",
                  "n",
                  "i32",
                ],
              ],
              "->",
              "i32",
              [
                "block",
                [
                  "if",
                  [
                    "<=",
                    "n",
                    1,
                  ],
                  [
                    ":",
                    "then",
                    [
                      "block",
                      "n",
                    ],
                  ],
                  [
                    ":",
                    "else",
                    [
                      "block",
                      [
                        "+",
                        [
                          "fib",
                          [
                            "-",
                            "n",
                            1,
                          ],
                        ],
                        [
                          "fib",
                          [
                            "-",
                            "n",
                            2,
                          ],
                        ],
                      ],
                    ],
                  ],
                ],
              ],
            ],
            [
              "fn",
              [
                "main",
              ],
              [
                "block",
                [
                  "let",
                  [
                    "=",
                    "x",
                    [
                      "+",
                      10,
                      [
                        "block",
                        [
                          "+",
                          20,
                          30,
                        ],
                      ],
                    ],
                  ],
                ],
                [
                  "let",
                  [
                    "=",
                    "y",
                    [
                      "if",
                      [
                        ">",
                        "x",
                        10,
                      ],
                      [
                        ":",
                        "then",
                        [
                          "block",
                          10,
                        ],
                      ],
                      [
                        ":",
                        "else",
                        [
                          "block",
                          20,
                        ],
                      ],
                    ],
                  ],
                ],
                [
                  "call",
                  "this",
                  "while",
                  [
                    "=>",
                    [],
                    [
                      "if",
                      [
                        ">",
                        "x",
                        10,
                      ],
                      [
                        ":",
                        "then",
                        [
                          "block",
                          [
                            "-=",
                            "x",
                            1,
                          ],
                        ],
                      ],
                      [
                        ":",
                        "else",
                        [
                          "block",
                          [
                            "+=",
                            "x",
                            1,
                          ],
                        ],
                      ],
                    ],
                  ],
                ],
                [
                  "let",
                  [
                    "=",
                    "n",
                    [
                      "if",
                      [
                        ">",
                        [
                          "len",
                          "args",
                        ],
                        1,
                      ],
                      [
                        ":",
                        "then",
                        [
                          "block",
                          [
                            "log",
                            "console",
                            [
                              "string",
                              "Hey there!",
                            ],
                          ],
                          [
                            "unwrap",
                            [
                              "parseInt",
                              [
                                "at",
                                "args",
                                1,
                              ],
                            ],
                          ],
                        ],
                      ],
                      [
                        ":",
                        "else",
                        [
                          "block",
                          10,
                        ],
                      ],
                    ],
                  ],
                ],
                [
                  "let",
                  [
                    "=",
                    "x2",
                    10,
                  ],
                ],
                [
                  "let",
                  [
                    "=",
                    "z",
                    [
                      "nothing",
                    ],
                  ],
                ],
                [
                  "let",
                  [
                    "=",
                    "test_spacing",
                    [
                      "fib",
                      "n",
                    ],
                  ],
                ],
                [
                  "let",
                  [
                    "=",
                    "result",
                    [
                      "fib",
                      "n",
                    ],
                  ],
                ],
              ],
            ],
          ],
        ],
      ],
    ],
    [
      "module",
      "std",
      [
        "exports",
        [],
      ],
      [
        [
          "use",
          [
            "::",
            "root",
            "all",
          ],
        ],
        [
          "pub",
          [
            "use",
            [
              "::",
              "macros",
              "all",
            ],
          ],
        ],
        [
          "module",
          "deep",
          [
            "exports",
            [],
          ],
          [
            [
              "use",
              [
                "::",
                "root",
                "all",
              ],
            ],
            [
              "pub",
              [
                "use",
                [
                  "::",
                  "hey",
                  "all",
                ],
              ],
            ],
            [
              "pub",
              "fn",
              [
                "deeply",
                [
                  ":",
                  "src",
                  "i32",
                ],
                [
                  ":",
                  "dest",
                  "i32",
                ],
              ],
              "->",
              "i32",
              [
                "block",
              ],
            ],
            [
              "module",
              "nested",
              [
                "exports",
                [],
              ],
              [
                [
                  "pub",
                  [
                    "use",
                    [
                      "::",
                      "hey",
                      "all",
                    ],
                  ],
                ],
                [
                  "pub",
                  "fn",
                  [
                    "nested",
                    [
                      ":",
                      "src",
                      "i32",
                    ],
                    [
                      ":",
                      "dest",
                      "i32",
                    ],
                  ],
                  "->",
                  "i32",
                  [
                    "block",
                  ],
                ],
                [
                  "use",
                  [
                    "::",
                    "root",
                    "all",
                  ],
                ],
                [
                  "module",
                  "hey",
                  [
                    "exports",
                    [],
                  ],
                  [
                    [
                      "use",
                      [
                        "::",
                        "root",
                        "all",
                      ],
                    ],
                    [
                      "use",
                      [
                        "::",
                        [
                          "::",
                          "super",
                          "macros",
                        ],
                        "all",
                      ],
                    ],
                    [
                      "pub",
                      "fn",
                      [
                        "hey",
                        [
                          ":",
                          "src",
                          "i32",
                        ],
                        [
                          ":",
                          "dest",
                          "i32",
                        ],
                      ],
                      "->",
                      "i32",
                      [
                        "block",
                      ],
                    ],
                  ],
                ],
              ],
            ],
          ],
        ],
        [
          "module",
          "memory",
          [
            "exports",
            [],
          ],
          [
            [
              "use",
              [
                "::",
                "root",
                "all",
              ],
            ],
            [
              "use",
              [
                "::",
                [
                  "::",
                  "super",
                  "macros",
                ],
                "all",
              ],
            ],
            [
              "global",
              "let",
              [
                "=",
                "header-size",
                8,
              ],
            ],
            [
              "global",
              "let",
              [
                "=",
                "size-index",
                0,
              ],
            ],
            [
              "global",
              "let",
              [
                "=",
                "type-index",
                4,
              ],
            ],
            [
              "global",
              "var",
              [
                "=",
                "stack-pointer",
                0,
              ],
            ],
            [
              "pub",
              "fn",
              [
                "copy",
                [
                  ":",
                  "src",
                  "i32",
                ],
                [
                  ":",
                  "dest",
                  "i32",
                ],
              ],
              "->",
              "i32",
              [
                "block",
                [
                  "bnr",
                  [
                    "memory",
                    "copy",
                    "voyd",
                  ],
                  [
                    "dest",
                    "src",
                    [
                      "size",
                      "src",
                    ],
                  ],
                ],
                "dest",
              ],
            ],
          ],
        ],
      ],
    ],
  ],
]
`;
